File: mobile/lib/src/models/service_status.dart
/// Mirrors ServiceStatusResponse from app/schemas/service_status.py.
///
/// The truthful "is my service good, and when does it lapse" view. Expiry is
/// status/balance-driven, not date-driven: prepaid lapses on balance exhaustion
/// (balance + grace/deactivation), postpaid only via dunning on overdue
/// invoices. `nextChargeAt` is the next charge/invoice date, never an expiry.
class ServiceStatus {
  ServiceStatus({
    required this.billingMode,
    required this.currency,
    required this.lowBalance,
    required this.inDunning,
    required this.services,
    this.primaryAction,
    this.balance,
    this.minBalance,
    this.graceUntil,
    this.deactivationAt,
    this.outstanding,
    this.oldestOverdueDueAt,
  });

  final String billingMode; // prepaid | postpaid
  final String currency;

  // Prepaid health (null for postpaid).
  final double? balance;
  final double? minBalance;
  final bool lowBalance;
  final DateTime? graceUntil;
  final DateTime? deactivationAt;

  // Postpaid health (null/false for prepaid).
  final double? outstanding;
  final DateTime? oldestOverdueDueAt;
  final bool inDunning;

  final List<ServiceStatusItem> services;
  final ServiceStatusAction? primaryAction;

  bool get isPrepaid => billingMode == 'prepaid';

  /// An active service is heading for a cut and the server says a financial
  /// action can prevent it. The client never derives this from subscription
  /// status or invoice rows.
  bool get needsRenewal =>
      services.any((s) => s.usable && (s.action?.isFinancial ?? false));

  ServiceStatusItem? forSubscription(String subscriptionId) {
    for (final service in services) {
      if (service.subscriptionId == subscriptionId) return service;
    }
    return null;
  }

  List<ServiceStatusItem> get unavailableServices => services
      .where((s) =>
          !s.usable &&
          const {'blocked', 'suspended', 'stopped'}.contains(s.status))
      .toList(growable: false);

  factory ServiceStatus.fromJson(Map<String, dynamic> json) {
    final list = (json['services'] as List?) ?? const [];
    return ServiceStatus(
      billingMode: json['billing_mode'] as String? ?? 'prepaid',
      currency: json['currency'] as String? ?? 'NGN',
      balance: _toDouble(json['balance']),
      minBalance: _toDouble(json['min_balance']),
      lowBalance: json['low_balance'] as bool? ?? false,
      graceUntil: _toDate(json['grace_until']),
      deactivationAt: _toDate(json['deactivation_at']),
      outstanding: _toDouble(json['outstanding']),
      oldestOverdueDueAt: _toDate(json['oldest_overdue_due_at']),
      inDunning: json['in_dunning'] as bool? ?? false,
      services: list
          .whereType<Map>()
          .map((e) => ServiceStatusItem.fromJson(e.cast<String, dynamic>()))
          .toList(),
      primaryAction: json['primary_action'] is Map
          ? ServiceStatusAction.fromJson(
              (json['primary_action'] as Map).cast<String, dynamic>())
          : null,
    );
  }
}

class ServiceStatusItem {
  ServiceStatusItem({
    required this.subscriptionId,
    required this.status,
    required this.billingMode,
    required this.usable,
    required this.reason,
    this.offerName,
    this.expiresAt,
    this.nextChargeAt,
    this.action,
  });

  final String subscriptionId;
  final String? offerName;
  final String status;
  final String billingMode;
  final bool usable;
  final DateTime? expiresAt;
  final DateTime? nextChargeAt;

  /// Server-owned service reason; clients do not infer actions from this text.
  final String reason;
  final ServiceStatusAction? action;

  bool get actionable => action != null;

  factory ServiceStatusItem.fromJson(Map<String, dynamic> json) =>
      ServiceStatusItem(
        subscriptionId: json['subscription_id'].toString(),
        offerName: json['offer_name'] as String?,
        status: json['status'] as String? ?? 'active',
        billingMode: json['billing_mode'] as String? ?? 'prepaid',
        usable: json['usable'] as bool? ?? false,
        expiresAt: _toDate(json['expires_at']),
        nextChargeAt: _toDate(json['next_charge_at']),
        reason: json['reason'] as String? ?? 'ok',
        action: json['action'] is Map
            ? ServiceStatusAction.fromJson(
                (json['action'] as Map).cast<String, dynamic>())
            : null,
      );
}

class ServiceStatusAction {
  ServiceStatusAction({
    required this.kind,
    required this.label,
    required this.message,
    required this.currency,
    required this.restoresService,
    this.amount,
  });

  /// top_up | pay_invoices | view_usage | contact_support
  final String kind;
  final String label;
  final String message;
  final double? amount;
  final String currency;

  /// True only when the server has proven this action clears every known hold.
  final bool restoresService;

  bool get isFinancial => kind == 'top_up' || kind == 'pay_invoices';

  factory ServiceStatusAction.fromJson(Map<String, dynamic> json) =>
      ServiceStatusAction(
        kind: json['kind'] as String? ?? 'contact_support',
        label: json['label'] as String? ?? 'Contact support',
        message: json['message'] as String? ?? 'Contact support for help.',
        amount: _toDouble(json['amount']),
        currency: json['currency'] as String? ?? 'NGN',
        restoresService: json['restores_service'] as bool? ?? false,
      );
}

double? _toDouble(dynamic v) {
  if (v == null) return null;
  if (v is num) return v.toDouble();
  return double.tryParse(v.toString());
}

DateTime? _toDate(dynamic v) {
  if (v == null) return null;
  return DateTime.tryParse(v.toString())?.toLocal();
}

File: app/schemas/service_status.py
"""Customer service-status schema — the single truthful "is my service good,
and when (if ever) does it lapse" view.

Service expiry in this system is NOT date-driven: prepaid lapses on balance
exhaustion (consumption-driven, surfaced here via balance + grace/deactivation
timers), postpaid lapses only via dunning on overdue invoices. `next_charge_at`
is the next charge/invoice date, never an expiry.
"""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from enum import StrEnum
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.status_presentation import StatusPresentation


class ServiceStatusActionKind(StrEnum):
    """Semantic customer action; clients only map it to their own route."""

    top_up = "top_up"
    pay_invoices = "pay_invoices"
    view_usage = "view_usage"
    contact_support = "contact_support"


class ServiceStatusAction(BaseModel):
    """Server-owned action hint derived from access and financial policy."""

    kind: ServiceStatusActionKind
    label: str
    message: str
    amount: Decimal | None = None
    currency: str = "NGN"
    # True only when completing this action clears every known access hold.
    restores_service: bool = False


class ServiceStatusItem(BaseModel):
    """Per-subscription truthful status for the caller's current services."""

    subscription_id: UUID
    offer_name: str | None = None
    status: str
    status_presentation: StatusPresentation
    billing_mode: str
    # Currently providing service (RADIUS/connectivity allowed).
    usable: bool
    # The date the service genuinely lapses if nothing changes, or null when it
    # has none. Never derived from next_charge_at.
    expires_at: datetime | None = None
    # Next billing event. Informational — NOT an expiry.
    next_charge_at: datetime | None = None
    # ok | low_balance | overdue | fair_usage | administrative_hold |
    # customer_hold | fraud_review | system_hold | multiple_holds | suspended |
    # stopped | ended
    reason: str
    action: ServiceStatusAction | None = None


class ServiceStatusResponse(BaseModel):
    """Account-level billing health plus per-service status."""

    as_of: datetime
    billing_mode: str
    currency: str = "NGN"

    # Prepaid health (null for postpaid accounts).
    balance: Decimal | None = None
    min_balance: Decimal | None = None
    low_balance: bool = False
    grace_until: datetime | None = None
    deactivation_at: datetime | None = None

    # Postpaid health (null/false for prepaid accounts).
    outstanding: Decimal | None = None
    oldest_overdue_due_at: datetime | None = None
    in_dunning: bool = False

    services: list[ServiceStatusItem] = Field(default_factory=list)
    # One account-level action for compact surfaces such as the mobile
    # dashboard. It prioritizes unavailable services over preventative nudges.
    primary_action: ServiceStatusAction | None = None
